File: app/src/test/java/com/tasomaniac/openwith/resolver/ChooserHistoryTest.java
package com.tasomaniac.openwith.resolver;

import android.content.SharedPreferences;

import org.junit.Rule;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;

import static junit.framework.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.BDDMockito.given;

public class ChooserHistoryTest {

    @Rule public MockitoRule rule = MockitoJUnit.rule();

    @Mock SharedPreferences preferences;

    @Test
    public void fromSettings() {
        String anySaveString = "key1|5#key2|3";
        givenSaveString(anySaveString);

        ChooserHistory history = ChooserHistory.fromSettings(preferences);

        assertEquals(anySaveString, history.getAsSaveString());
    }

    @Test(expected = NumberFormatException.class)
    public void shouldThrowExceptionForInvalid() {
        String anySaveString = "key1|5#key2|a";
        givenSaveString(anySaveString);

        ChooserHistory history = ChooserHistory.fromSettings(preferences);

        assertEquals(anySaveString, history.getAsSaveString());
    }

    @Test
    public void shouldOmitInvalidParts() {
        givenSaveString("key1|5#key2#key3|4");

        ChooserHistory history = ChooserHistory.fromSettings(preferences);

        assertEquals("key1|5#key3|4", history.getAsSaveString());
    }

    @Test
    public void shouldOmitInvalidLastPart() {
        givenSaveString("key1|5#key2|5#key3");

        ChooserHistory history = ChooserHistory.fromSettings(preferences);

        assertEquals("key1|5#key2|5", history.getAsSaveString());
    }

    private void givenSaveString(String anySaveString) {
        given(preferences.getString(anyString(), anyString())).willReturn(anySaveString);
    }

}

File: resolver/src/test/kotlin/com/tasomaniac/openwith/resolver/ChooserHistoryTest.kt
package com.tasomaniac.openwith.resolver

import androidx.core.content.edit
import com.tasomaniac.openwith.TestSharedPreferences
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class ChooserHistoryTest {

    private val preferences = TestSharedPreferences()

    @Test
    fun fromSettings() {
        val anySaveString = "key1|5#key2|3"
        save(anySaveString)

        val history = ChooserHistory.fromSettings(preferences)

        assertThat(history.asSaveString).isEqualTo(anySaveString)
    }

    @Test(expected = NumberFormatException::class)
    fun shouldThrowExceptionForInvalid() {
        val anySaveString = "key1|5#key2|a"
        save(anySaveString)

        val history = ChooserHistory.fromSettings(preferences)

        assertThat(history.asSaveString).isEqualTo(anySaveString)
    }

    @Test
    fun shouldOmitInvalidParts() {
        save("key1|5#key2#key3|4")

        val history = ChooserHistory.fromSettings(preferences)

        assertThat(history.asSaveString).isEqualTo("key1|5#key3|4")
    }

    @Test
    fun shouldOmitInvalidLastPart() {
        save("key1|5#key2|5#key3")

        val history = ChooserHistory.fromSettings(preferences)

        assertThat(history.asSaveString).isEqualTo("key1|5#key2|5")
    }

    private fun save(value: String) {
        preferences.edit {
            putString("history", value)
        }
    }

}
